validate/validate.go

```diff
@@ -0,0 +1,574 @@
+		"CAP_SETFCAP",
+		"CAP_SETPCAP",
+		"CAP_NET_BIND_SERVICE",
+		"CAP_SYS_CHROOT",
+		"CAP_KILL",
+		"CAP_AUDIT_WRITE",
+	}
+)
+
+type Validator struct {
+	spec         rspec.Spec
+	rootfs       string
+	hostSpecific bool
+}
+
+func NewValidator(spec rspec.Spec, rootfs string, hostSpecific bool) Validator {
+	return Validator{spec: spec, rootfs: rootfs, hostSpecific: hostSpecific}
+}
+
+func (v *Validator) CheckAll() (msgs []string) {
+	msgs = append(msgs, v.CheckMandatoryFields()...)
+	msgs = append(msgs, v.CheckSemVer()...)
+	msgs = append(msgs, v.CheckMounts()...)
+	msgs = append(msgs, v.CheckPlatform()...)
+	msgs = append(msgs, v.CheckProcess()...)
+	msgs = append(msgs, v.CheckLinux()...)
+	msgs = append(msgs, v.CheckHooks()...)
+
+	return
+}
+
+func (v *Validator) CheckSemVer() (msgs []string) {
+	logrus.Debugf("check semver")
+
+	version := v.spec.Version
+	_, err := semver.Parse(version)
+	if err != nil {
+		msgs = append(msgs, fmt.Sprintf("%q is not valid SemVer: %s", version, err.Error()))
+	}
+	if version != rspec.Version {
+		msgs = append(msgs, fmt.Sprintf("internal error: validate currently only handles version %s, but the supplied configuration targets %s", rspec.Version, version))
+	}
+
+	return
+}
+
+func (v *Validator) CheckPlatform() (msgs []string) {
+	logrus.Debugf("check platform")
+
+	validCombins := map[string][]string{
+		"darwin":    {"386", "amd64", "arm", "arm64"},
+		"dragonfly": {"amd64"},
+		"freebsd":   {"386", "amd64", "arm"},
+		"linux":     {"386", "amd64", "arm", "arm64", "ppc64", "ppc64le", "mips64", "mips64le"},
+		"netbsd":    {"386", "amd64", "arm"},
+		"openbsd":   {"386", "amd64", "arm"},
+		"plan9":     {"386", "amd64"},
+		"solaris":   {"amd64"},
+		"windows":   {"386", "amd64"}}
+	platform := v.spec.Platform
+	for os, archs := range validCombins {
+		if os == platform.OS {
+			for _, arch := range archs {
+				if arch == platform.Arch {
+					return nil
+				}
+			}
+			msgs = append(msgs, fmt.Sprintf("Combination of %q and %q is invalid.", platform.OS, platform.Arch))
+		}
+	}
+	msgs = append(msgs, fmt.Sprintf("Operation system %q of the bundle is not supported yet.", platform.OS))
+
+	return
+}
+
+func checkEventHooks(hookType string, hooks []rspec.Hook, hostSpecific bool) (msgs []string) {
+	for _, hook := range hooks {
+		if !filepath.IsAbs(hook.Path) {
+			msgs = append(msgs, fmt.Sprintf("The %s hook %v: is not absolute path", hookType, hook.Path))
+		}
+
+		if hostSpecific {
+			fi, err := os.Stat(hook.Path)
+			if err != nil {
+				msgs = append(msgs, fmt.Sprintf("Cannot find %s hook: %v", hookType, hook.Path))
+			}
+			if fi.Mode()&0111 == 0 {
+				msgs = append(msgs, fmt.Sprintf("The %s hook %v: is not executable", hookType, hook.Path))
+			}
+		}
+
+		for _, env := range hook.Env {
+			if !envValid(env) {
+				msgs = append(msgs, fmt.Sprintf("Env %q for hook %v is in the invalid form.", env, hook.Path))
+			}
+		}
+	}
+
+	return
+}
+func (v *Validator) CheckHooks() (msgs []string) {
```

Comment:
There's unnecessary churn in:

    $ diff -u <(git cat-file -p origin/pr/188^^:cmd/ocitools/validate.go) <(git cat-file -p origin/pr/188^:validate/validate.go)

because you shifted this from being before `checkEventHooks` to being after it.  Can we move it back up?